analytics/input_redactor.go

```diff
@@ -0,0 +1,274 @@
+package analytics
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"reflect"
+	"sort"
+
+	"github.com/bitrise-io/bitrise/bitrise"
+	bitriseModels "github.com/bitrise-io/bitrise/models"
+	"github.com/bitrise-io/bitrise/tools/filterwriter"
+	"github.com/bitrise-io/envman/env"
+	"github.com/bitrise-io/envman/models"
+)
+
+// InputRedactor ...
+type InputRedactor interface {
+	Redact(inputs []models.EnvironmentItemModel) (map[string]interface{}, error)
+}
+
+type inputRedactor struct {
+	secrets           []string
+	environment       []models.EnvironmentItemModel
+	buildRunResults   bitriseModels.BuildRunResultsModel
+	isCIMode          bool
+	isPullRequestMode bool
+	envSource         env.EnvironmentSource
+}
+
+// NewInputRedactor ...
+func NewInputRedactor(secrets []string, environment []models.EnvironmentItemModel, buildRunResults bitriseModels.BuildRunResultsModel, isCIMode bool, isPullRequestMode bool, envSource env.EnvironmentSource) InputRedactor {
+	return inputRedactor{secrets: secrets, environment: environment, buildRunResults: buildRunResults, isCIMode: isCIMode, isPullRequestMode: isPullRequestMode, envSource: envSource}
+}
+
+// Redact ...
+func (i inputRedactor) Redact(inputs []models.EnvironmentItemModel) (map[string]interface{}, error) {
+	result := map[string]interface{}{}
+	for _, input := range inputs {
+		options, key, value, err := getData(input)
+		if err != nil {
+			return nil, err
+		}
+		raw := map[string]interface{}{}
+		expanded := map[string]interface{}{}
+		if err := i.fillMap(options, raw, expanded, key, reflect.ValueOf(value)); err != nil {
+			return nil, err
+		}
+		var finalValue interface{}
+		if v, ok := expanded[key]; ok {
+			finalValue = v
+		} else {
+			finalValue = raw[key]
+		}
+		valueMap := map[string]interface{}{"value": finalValue}
+		result[key] = valueMap
+		if !reflect.DeepEqual(raw[key], expanded[key]) {
+			valueMap["original_value"] = raw[key]
+		}
+	}
+	return result, nil
+}
+
+func (i inputRedactor) growSlice(options models.EnvironmentItemOptionsModel, container []interface{}, expandedContainer []interface{}, value reflect.Value) ([]interface{}, []interface{}, error) {
```

Comment:
Supporting maps and slices might be an overshoot now (however it would be great to support maps and arrays for input values)